README.md
# RCT
RCT helps you focus on the statistics of the randomized control trials, rather than the heavy programming lifting. Any social scientist should use RCT when conducting Field Experiments.

RCT helps you in the whole process of designing and evaluating a RCT. 

Let the steps of the design be a RCT: 

### 1. Get data in which you want to randomly assign treatment

This library has a function called **summary_statistics** to know the distribution of all covariates in data. 

### 2. Decide the share of observations that will go to control group 

The function **tau_min** calculates the minimum detectable treatment effect given power, significance level, outcome variable and number of observations.  This function computes this for any share of control observations. 

### 3. Decide which variables to use for strata building

Prior to random assignment, one has to decide which **categorical** variables to build blocks. Hence, the blocks or strata are the group that combine every categorical variable. The cardinality of this groups are all the possible combinations of the chose categorical variables. 

To build categorical variables in a powerful way, function **n_tile_label** divides variables in the decided n groups, putting label of the range of each category to the variable. 

### 4. Random Assignment 

Once we have the blocking variables, we need to assign treatment status **within** each strata. Function treatment_assign performs such random assignment for any given number of treatment groups. Furthermore, it handles misfits. 

Misfits are defined as observations within each strata that are not really randomly assigned because when dividing the size of each strata N_strata to each treatment share, there are some remainder observations. 

For instance, let the following example: 

N_strata = 10 
share_control = $\frac{1}{3}$
share_treat_1 = $\frac{1}{3}$
share_treat_2 = $\frac{1}{3}$

First 3 units are assigned to control, second 3 units are assigned to treat 1 and the last 3 unit are assigned to treat 2. As you already notice, the last observation is the remainder. This is a misfit. Misfits alter the successful random assignment because they are not. In the example, this observation is assigned to treat 2 non-randomly. 

The function **treatment_assign** handles misfits in three ways. 

- "NA" assigns the misfits to NAs, leaving the experiment with only the pure assigned observations. 

- "global" puts together the misfits of each strata into a single group and then assigns them randomly

- "strata" assigns misfits to treatment within each strata

### 5 Impact evaluation 

After running a RCT, the social scientist wants to know the ATE for one or several variables and the distribution of this impact within the blocking variables to check for Heterogenous Treatment Effects. Additionally, if the experiment lasted for more than one period and panel-data is available, one must cluster the standard errors by each i unit and control for period fixed effects. Finally, if by chance one o more covariates are not balance, one would like to control for them. 

Function **impact_eval** does all this jobs in one single command. It runs all the ATE regressions for each endogenous variable, all the combinations of endogenous variables*heterogenous variables. 

For each combination the model run is:

$$Y_i = \alpha + \tau treatment + \epsilon $$

<!-- badges: start -->
[![Travis build status](https://travis-ci.org/isidorogu/RCT.svg?branch=master)](https://travis-ci.org/isidorogu/RCT)
<!-- badges: end -->


NEWS.md
# RCT 1.0.2
RCT helps you focus on the statistics of the randomized control trials, rather than the heavy programming lifting. Any social scientist should use RCT when conducting Field Experiments.

RCT helps you in the whole process of designing and evaluating a RCT. 

This version ensures that all functions remain compatible with upcoming dplyr 1.0.0
- balance_regression should continue to work well in both dplyr 0.89 and dplyr 1.0.0

# RCT 1.0.2.1

This version ensures that all functions remain compatible with broom 0.70

tests/testthat.R
library(testthat)
library(RCT)

test_check("RCT")


tests/testthat/test_tau_min.R
context("Pruebas tau_min/tau_min_probability")

#' @importFrom magrittr %>%

# Base completa para TODOS los tests 
data_original<-data.frame(key = c(1:1000), 
                          inc_quartile = rep(c("Q1", "Q2", "Q3", "Q4"), each = 250), 
                          age_quartile = rep(c("Q1", "Q2", "Q3", "Q4"), times = 250), 
                          inc = rnorm(n = 1000, mean = 8000, sd = 4000), 
                          age = rpois(n = 1000, lambda = 34), 
                          log_outcome = rnorm(n = 1000, mean = 8, sd = 1.4), 
                          treat = rep(c(rep(c(0,1,2,3), times = 10), rep(c(0,1,2,3), each = 10)), each = 25)) 

# 1. Dimensiones 
a<-tau_min(data_original$log_outcome,
           N = nrow(data_original), share_control = seq(0.05, 0.95, 0.05))

b<-tau_min(data_original %>% dplyr::select(log_outcome),
           N = nrow(data_original), share_control = seq(0.05, 0.95, 0.05))

e<-tau_min_probability(prior = 0.5, N = nrow(data_original), share_control =  seq(0.05, 0.95, 0.05))


test_that("dimensions tau_min", {
  expect_is(a, "data.frame")
  expect_is(b, "data.frame")
  expect_equal(ncol(a), 9)
  expect_equal(ncol(b), 9)
  expect_equal(ncol(e), 9)
  expect_equal(nrow(a), 19)
  expect_equal(nrow(b), 19)
  expect_equal(nrow(e), 19)
  expect_equal(a,b)
  
})




# Errores
test_that("Errors in N or outcome", {
expect_error(tau_min_probability("c", nrow(diamonds), share_control = seq(0.05,0.95, 0.05)))
expect_error(tau_min(diamonds$depth, N = 1.5, share_control = seq(0.05, 0.95, 0.05)))
})




tests/testthat/test_impact_eval.R
context("Pruebas impact_eval")

#' @importFrom magrittr %>%

# Base completa para TODOS los tests 
data_original<-data.frame(key = c(1:1000), 
                          inc_quartile = rep(c("Q1", "Q2", "Q3", "Q4"), each = 250), 
                          age_quartile = rep(c("Q1", "Q2", "Q3", "Q4"), times = 250), 
                          inc = rnorm(n = 1000, mean = 8000, sd = 4000), 
                          age = rpois(n = 1000, lambda = 34), 
                          log_outcome = rnorm(n = 1000, mean = 8, sd = 1.4), 
                          treat = rep(c(rep(c(0,1,2,3), times = 10), rep(c(0,1,2,3), each = 10)), each = 25)) 


# 1. Dimensiones 
imp_eval<-impact_eval(data = data_original, 
                      endogenous_vars = "log_outcome", treatment = "treat")

test_that("impact_eval correct dimensions tests", {
  expect_is(imp_eval, "list")
  expect_equal(length(imp_eval), 1)
  expect_equal(ncol(imp_eval$log_outcome), 5)
  expect_equal(nrow(imp_eval$log_outcome), 4)

  
})


# 2. Nombres
test_that("Correct names impact_eval" , {
  expect_equal(names(imp_eval$log_outcome), c("term", "estimate", "std.error", "statistic", "p.value"))
  expect_equal(sum(stringr::str_detect(unique(imp_eval$log_outcome$term), pattern = "Intercept")), 1)
})



####
# Dos variables

imp_eval2<-impact_eval(data = data_original, 
                      endogenous_vars = c("log_outcome", "inc"), treatment = "treat")

test_that("impact_eval2 correct dimensions tests", {
  expect_is(imp_eval2, "list")
  expect_equal(length(imp_eval2), 2)
  expect_equal(ncol(imp_eval2$log_outcome), 5)
  expect_equal(ncol(imp_eval2$inc), 5)
  expect_equal(nrow(imp_eval2$log_outcome), 4)
  expect_equal(nrow(imp_eval2$inc), 4)

})

# 2. Nombres
test_that("Correct names impact_eval2" , {
  expect_equal(names(imp_eval2$log_outcome), c("term", "estimate", "std.error", "statistic", "p.value"))
  expect_equal(sum(stringr::str_detect(unique(imp_eval2$log_outcome$term), pattern = "Intercept")), 1)
  expect_equal(sum(stringr::str_detect(unique(imp_eval2$inc$term), pattern = "Intercept")), 1)
})




# Con heterogeneidades, fixed effects y controles 
imp_eval3<-impact_eval(data = data_original, 
                      endogenous_vars = "log_outcome", 
                      treatment = "treat", 
                      heterogenous_vars = "inc_quartile", 
                      fixed_effect_vars = "age_quartile", control_vars = "inc")

test_that("impact_eval correct dimensions tests", {
  expect_is(imp_eval3, "list")
  expect_equal(length(imp_eval3), 2)
  expect_equal(ncol(imp_eval3$log_outcome), 5)
  expect_equal(nrow(imp_eval3$log_outcome), 4)
  expect_equal(ncol(imp_eval3$log_outcome_inc_quartile), 6)
  expect_equal(nrow(imp_eval3$log_outcome_inc_quartile), 16)
  
  
})


# 2. Nombres
test_that("Correct names impact_eval" , {
  expect_equal(names(imp_eval3$log_outcome), c("term", "estimate", "std.error", "statistic", "p.value"))
  expect_equal(sum(stringr::str_detect(unique(imp_eval3$log_outcome$term), pattern = "Intercept")), 0)
  expect_equal(sum(stringr::str_detect(unique(imp_eval3$log_outcome_inc_quartile$term), pattern = "Intercept")), 0)
  
})




R/RCT.R
#' Designing, random assigning and evaluating Randomized Control Trials
#' 
#' RCT provides three important group of functions:
#' a) functions for pre-processing the design of the RCT 

#' b) Functions for assigning treatment status and checking for balances 

#' c) Function for evaluating the impact of the RCT
#' 
#' @docType package
#' @name RCT
#' @details 
#' RCT helps you focus on the statistics of the randomized control trials, 
#' rather than the heavy programming lifting. 
#' RCT helps you in the whole process of designing and evaluating a RCT. 
#' 1. Clean and summarise the data in which you want to randomly assign treatment 
#' 2. Decide the share of observations that will go to control group 
#' 3. Decide which variables to use for strata building
#' 4. Robust Random Assignment by strata/blocks
#' 5 Impact evaluation of all y's and heterogeneities
#' To lean more about RCT, start with the vignette: browseVignettes(package = "RCT")
#' @section RCT functions: 
#' treatment_assign: Robust treatment assign by strata/blocks
#' 
#' impact_eval: Automatized impact evaluation with heterogeneous treatment effects
#' 
#' balance_table: Balance tables for any length of covariates
#' 
#' balance_regression: LPM of treatment status against covariates with F-test
#' 
#' tau_min: Computation of the minimum detectable effect between control & treatment units
#' 
#' tau_min_probability: Computation of the minimum detectable effect between control & treatment units for dicotomical y-vars
#' 
#' summary_statistics: Summary statistics of all numeric columns in your data
#' 
#' ntile_label: Rank and divide observations in n groups, with label
#' @references 
#' Athey, Susan, and Guido W. Imbens (2017) "The Econometrics Randomized Experiments". Handbook of economic field experiments. \url{https://arxiv.org/abs/1607.00698}
#' @seealso   
#' Useful links:
#' https://github.com/isidorogu/RCT
#' Report bugs at https://github.com/isidorogu/RCT/issues

#' @author Isidoro Garcia Urquieta, isidoro.gu@@gmail.com

NULL


R/tau_min.R
#' tau_min() computes the minimum detectable difference between control group and each treatment
#' @param outcome_var the variable for which you wish to test the impact of treatment 
#' @param N number of observations in the RCT, usually nrow(data)
#' @param significance The level of significance of the test Pr(Reject H_0 | H_0 False). Default is 0.05
#' @param power The level of power of the test (1 - Pr(Reject H_0 | H_0 True) ). Default is 0.8
#' @param share_control The share of observations in N assigned to control. This argument allows for sequences (i.e. seq(0,1,0.1))
#' @param n_groups Number of groups (control + # treatment groups)
#' @return A tibble with the share_control and N observations in control group (N_control), 
#' the share and N of each treatment c(share_ti, N_ti), 
#' total share of treatment rows and N treated (share_treat, N_treat), N, 
#' the minimum detectable difference between control and all treatments together (tau_min_global),
#' the minimum detectable difference between control and each treatment (tau_min_each_treat)
#' @examples 
#'data <- data.frame(y_1 = rbinom(n = 100, size = 1, prob = 0.3), 
#'                   y_2 = rnorm(n = 100, mean = 8, sd = 2))
#' tau_min(data$y_1, N = nrow(data), share_control = seq(0,1,0.1), n_groups = 3)
#' @details This function calculates the minimum difference that could show significant 
#' E[Y(1)-Y(0)] = tau, between any two given groups (e.g. control vs each treatment), given the 
#' population size (N), the outcome variable, power and significance

#' @export
#' @importFrom magrittr %>%
tau_min <- function(outcome_var,
                    N,
                    power = 0.8,
                    significance = 0.05,
                    share_control, n_groups = 2) {
    if (!(power > 0 & power < 1 & significance > 0 & significance < 1)) {
        stop("power and significan must be numbers between 0 and 1")
    }
    else {

    if (!("integer" %in% class(N))){
        stop("N must be integer")
    }
    else if (N < 0) {
        stop("N must be postive")
    }
    else if ("data.frame" %in% class(outcome_var) ){


            outcome_var <- dplyr::pull(outcome_var)

            share_ti <- (1-share_control)/(n_groups - 1)

            N_each_treat_comp <- N*share_control+N*share_ti

            share_control_ti <- share_control/(share_control + share_ti)

            variance <- stats::var(outcome_var, na.rm = T)

            estadistico <- (stats::qnorm(power) + stats::qnorm(1 - significance/2))^2

            tau_min_global <-  (estadistico*variance)/(N * share_control * (1 - share_control))

            tau_min_each_trat <- tau_min_global*(N * share_control * (1 - share_control))/
                (N_each_treat_comp * share_control_ti * (1 - share_control_ti))


            tau_min_DT <- data.frame(share_control = share_control,
                                 share_ti = share_ti,
                                 share_treat = share_ti*(n_groups-1),
                                 N = N,
                                 N_control = N*share_control,
                                 N_treat = (n_groups-1)*share_ti*N,
                                 N_ti = share_ti*N,
                                 tau_min_global = sqrt(tau_min_global),
                                 tau_min_each_treat = sqrt(tau_min_each_trat))


            return(tau_min_DT)

    }  else {
        share_ti <- (1-share_control)/(n_groups - 1)
        
        N_each_treat_comp <- N*share_control+N*share_ti
        
        share_control_ti <- share_control/(share_control + share_ti)
        
        variance <- stats::var(outcome_var, na.rm = T)
        
        estadistico <- (stats::qnorm(power) + stats::qnorm(1 - significance/2))^2
        
        tau_min_global <-  (estadistico*variance)/(N * share_control * (1 - share_control))
        
        tau_min_each_trat <- tau_min_global*(N * share_control * (1 - share_control))/
            (N_each_treat_comp * share_control_ti * (1 - share_control_ti))
        
        
        tau_min_DT <- data.frame(share_control = share_control,
                                 share_ti = share_ti,
                                 share_treat = share_ti*(n_groups-1),
                                 N = N,
                                 N_control = N*share_control,
                                 N_treat = (n_groups-1)*share_ti*N,
                                 N_ti = share_ti*N,
                                 tau_min_global = sqrt(tau_min_global),
                                 tau_min_each_treat = sqrt(tau_min_each_trat))
        
        
        return(tau_min_DT)
        
    }
        
    }
    }
    
    
        
        



#' tau_min_probability() computes the minimum detectable difference between control group and each treatment for a dicotomical variable
#' @param prior Pr(Y=1). 
#' @param N number of observations in the RCT, usually nrow(data)
#' @param significance The level of significance of the test Pr(Reject H_0 | H_0 False). Default is 0.05
#' @param power The level of power of the test (1 - Pr(Reject H_0 | H_0 True) ). Default is 0.8
#' @param share_control The share of observations in N assigned to control. This argument allows for sequences (i.e. seq(0,1,0.1))
#' @param n_groups Number of groups (control + # treatment groups)
#' @return A tibble with the share_control and N observations in control group (N_control), 
#' the share and N of each treatment c(share_ti, N_ti), 
#' total share of treatment rows and N treated (share_treat, N_treat), N, 
#' the minimum detectable difference between control and all treatments together (tau_min_global),
#' the minimum detectable difference between control and each treatment (tau_min_each_treat)
#' @examples 
#' tau_min_probability(0.4, N = 1000, share_control = seq(0,1,0.1), n_groups = 3)
#' @details This function calculates the minimum difference that could show significant 
#' Pr[Y(1)-Y(0)] = tau, between any two given groups (e.g. control vs each treatment), given the 
#' population size (N), the outcome variable, power and significance
#' @export
#' @importFrom magrittr %>%
tau_min_probability <- function(prior,
                                N,
                                power = 0.8,
                                significance = 0.05,
                                share_control,
                                n_groups = 2) {
    if (!(prior>0 & prior <1)){
        stop("prior must be a number between 0 and 1")
    } else {

    share_ti = (1-share_control)/(n_groups - 1)

    N_each_treat_comp = N*share_control+N*share_ti

    share_control_ti = share_control/(share_control + share_ti)

    variance = prior*(1-prior)

    estadistico  = (stats::qnorm(power) + stats::qnorm(1 - significance/2))^2

    tau_min_global =  (estadistico*variance)/(N * share_control * (1 - share_control))

    tau_min_each_trat = tau_min_global*(N * share_control * (1 - share_control))/
        (N_each_treat_comp * share_control_ti * (1 - share_control_ti))


    tau_min_DT <- data.frame(share_control = share_control,
                         share_ti = share_ti,
                         share_treat = share_ti*(n_groups-1),
                         N = N,
                         N_control = N*share_control,
                         N_treat = (n_groups-1)*share_ti*N,
                         N_ti = share_ti*N,
                         tau_min_global = sqrt(tau_min_global),
                         tau_min_each_treat = sqrt(tau_min_each_trat))


    return(tau_min_DT)
    }
    }






R/impact_eval.R
#' Impact Evaluation of Treatment Effects
#' @param data A data.frame, tibble or data.table
#' @param endogenous_vars Vector of Y's on which treatment effects will be evaluated
#' @param treatment Variable indicating the treatment status
#' @param heterogenous_vars Vector of variables for which you wish to assess treatment distributions/heterogeneities.
#' @param cluster_vars Vector of variables to cluster the standard errors. Default is without clustered std errors
#' @param fixed_effect_vars Vector of variables to add as fixed effects. Default is without fixed effects
#' @param control_vars Vector of variables to control for in the evaluation. Default is without controls
#' @return impact_eval() returns a list of regression tables. The names of the list are the same as the endogenous variables.
#' for heterogeneities the names are endogenous_var_heterogenous_var
#' @examples 
#' data <- data.frame(y_1 = rnorm(n = 100, mean = 100, sd = 15), 
#'                   y_2 = rnorm(n = 100, mean = 8, sd = 2), 
#'                   treat = rep(c(0,1,2,3), each = 25), 
#'                   heterogenous_var1 = rep(c("X_Q1", "X_Q2", "X_Q3", "X_Q4"), times = 25),
#'                   cluster_var1 = rep(c(1:5), times = 20), 
#'                   fixed_effect_var1 = rep(c(1,2), times = 50),
#'                   control_var1 = rnorm(n = 100, mean = 20, sd = 1))
#'
#' evaluation<-impact_eval(data = data, 
#'                        endogenous_vars = c("y_1", "y_2"), 
#'                        treatment = "treat", 
#'                        heterogenous_vars = c("heterogenous_var1"), 
#'                        cluster_vars = "cluster_var1", fixed_effect_vars = c("fixed_effect_var1"), 
#'                        control_vars = c("control_var1"))
#' @details This function carries out the evaluation of treatment effects on endogenous variables. 
#' It automatically runs the regressions of all the endogenous_vars supplied & all the combinations 
#' of endogenous_vars and heterogenous_vars. Additionally, the function has the option of include 
#' fixed_effects, controls and cluster variables for clustered std errors.

#' @export
#' @importFrom magrittr %>%
impact_eval <- function(data, endogenous_vars, treatment, 
                         heterogenous_vars, cluster_vars = "0", 
                         fixed_effect_vars = "0", control_vars) {
  
  # Poniendo como sumas los fixed effects, cluster std errors, controles 
  cluster_formula<- stringr::str_c(cluster_vars, collapse = "+" )
  fixed_effect_formula<- stringr::str_c(fixed_effect_vars, collapse = "+")
  
  if (missing(control_vars)) {
    
    formula_sin_y <-stringr::str_c("{.}~factor({treatment}) ",
                                   " | ", 
                                   fixed_effect_formula, 
                                   " | ", 
                                   " 0 | ", 
                                   cluster_formula)
    
    formulas<-purrr::map_chr(endogenous_vars  , ~glue::glue(formula_sin_y))
    
    ITT <- purrr::map(formulas,  ~lfe::felm(stats::as.formula(.), data = data ) %>% broom::tidy(.) )
    
    base::names(ITT)<-endogenous_vars
    
    
  } else {
    
    controles_formula<- stringr::str_c(control_vars, collapse = "+")
    
   formula_sin_y <-stringr::str_c("{.}~factor({treatment}) + ", 
                       controles_formula, 
                       " | ", 
                       fixed_effect_formula, 
                       " | ", 
                       " 0 | ", 
                       cluster_formula)
  
  formulas<-purrr::map_chr(endogenous_vars  , ~glue::glue(formula_sin_y))

  ITT <- purrr::map(formulas,  ~lfe::felm(stats::as.formula(.), data = data ) %>% broom::tidy(.) )
  
  base::names(ITT)<-endogenous_vars
  
  }
  
  if (missing(heterogenous_vars)) { 
    
    return(ITT)
  } else {
      

  # Heterogeneidades: mejor con formulas
    
    matrix<-base::expand.grid(heterogenous_vars, endogenous_vars)
    
    Var1<-NULL
    Var2<-NULL
    .<-NULL
    fit<-NULL
    matrix<-
      matrix %>%
      dplyr::rename(variables_heterogeneas = Var1,
                    variables_endogenas = Var2)
    
    endogenous_vars_final<-base::as.character(matrix$variables_endogenas)
    heterogenous_vars_final<-base::as.character(matrix$variables_heterogeneas)
    

  formulas_het<-purrr::map_chr(endogenous_vars_final  , ~glue::glue(formula_sin_y))

  if (utils::packageVersion("broom") > "0.5.6") {
    
    ITT_het<-purrr::map2(heterogenous_vars_final, formulas_het,
                            function(x, y) data %>%
                              dplyr::group_by(!!rlang::sym(x)) %>% 
                              dplyr::do(fit = lfe::felm(stats::as.formula(y),
                                                        data = .)) ) %>%
      purrr::map(., function(x) purrr::map_dfr(x$fit, broom::tidy))
    
    ITT_het<-purrr::map2(heterogenous_vars_final, ITT_het, 
                         function(x,y) { 
                           
                           valores_het<-base::unique(dplyr::pull(data, !!rlang::sym(x) ) ) 
                           
                           y<-
                             y %>% 
                             dplyr::mutate(!!x  := rep(valores_het, each = nrow(y)/length(valores_het) ) ) %>%
                             dplyr::select(!!x, dplyr::everything())
                         }
    )
    
  
    base::names(ITT_het)<-stringr::str_c(endogenous_vars_final, heterogenous_vars_final, sep = "_")
  
    regresiones_final<-c(ITT, ITT_het)
    
    return(regresiones_final)
  }
  else {
    
    ITT_het<-purrr::map2(heterogenous_vars_final, formulas_het,
                         function(x, y) data %>%
                           dplyr::group_by(!!rlang::sym(x)) %>% 
                           dplyr::do(fit = lfe::felm(stats::as.formula(y),
                                                     data = .)) %>% broom::tidy(., fit)) 
    
    base::names(ITT_het)<-stringr::str_c(endogenous_vars_final, heterogenous_vars_final, sep = "_")
    
    regresiones_final<-c(ITT, ITT_het)
    
    return(regresiones_final)
    
  }
}

}


R/balance_table.R
#' Creates balance table for the X variables across treatment status
#' @param data A data.frame, tibble or data.table
#' @param treatment a string with treatment status column
#' @return A tibble with Mean_value of each treatment status and p_values
#' @examples 
#' data <-data.frame(x1 = rnorm(n = 100, mean = 100, sd = 15), 
#'                   x2=rnorm(n = 100, mean = 65), 
#'                   treatment = rep(c(0,1,2,3,4), each = 20))
#' balance_table(data, "treatment")
#' @details balance_table() performs t.test(X~treatment) for each X column in data. Every value of 
#' treatment i.e 1,2,3,...N is compared against control value (0) or the first value of the treatment
#' column. For instance, If treatment column has values of (0,1,2,3), balance_table will return: 
#' the mean value of each treatment (for all X's), and the p_values of the t.test of (1,2,3) against
#' treatment = 0.  

#' @export
#' @importFrom magrittr %>%
balance_table<-function(data, treatment) {
  
    variables<-NULL 
    
    data <- data %>% dplyr::arrange(!!rlang::sym(treatment))
    
    valores_trat <- base::unique(dplyr::pull(data, !!rlang::sym(treatment)))
    
    trats<-valores_trat[2:base::length(valores_trat)]
    
    bal_tables<-purrr::map(trats, function (x)
      data %>%
        dplyr::filter(!!rlang::sym(treatment) == valores_trat[1] | !!rlang::sym(treatment) ==  !!x))
    
    
    
    bal_tables<-purrr::map(bal_tables, function (x) x %>%
                             tidyr::pivot_longer(names_to = "variables", values_to = "value", -treatment) )
    
    
    # Creo por separado la primera para poner la media de control 
    
    bal_tables<-purrr::map(bal_tables, function(x) x %>% 
                             dplyr::group_by(variables) %>% 
                             dplyr::summarise(Media_control = stats::t.test(value~!!rlang::sym(treatment))$estimate[1],
                                              Media_trat = stats::t.test(value~!!rlang::sym(treatment))$estimate[2], 
                                              p_value = stats::t.test(value~!!rlang::sym(treatment))$p.value))
    
    valores_trat<-base::as.character(valores_trat[2:base::length(valores_trat)])
    
    bal_tables <- purrr::map2_dfc(.x = bal_tables, .y = valores_trat,
                                  function(x,y) dplyr::rename_all(x, ~stringr::str_c(., y)))
    
    #Quedandome solo con una de variables y media_control
    medias_control <- base::names(bal_tables %>% dplyr::select(dplyr::contains("control")) )
    variables_nombres <- base::names(bal_tables %>% dplyr::select(dplyr::contains("variables")))
    medias_trat <- base::names(bal_tables %>% dplyr::select(dplyr::contains("trat")))
    p_values <-base::names(bal_tables %>% dplyr::select(dplyr::contains("p_value")))
    
    bal_tables<-
      bal_tables %>% 
      dplyr::select(tidyselect::all_of(variables_nombres[1]), 
                    tidyselect::all_of(medias_control[1]), 
                    tidyselect::all_of(medias_trat), 
                    tidyselect::all_of(p_values))
    
    return(bal_tables)
}




R/balance_regression.R
#' balance_regression() Runs a LPM of treatment status against all covariates (treatment~X'B). 
#' @param data A data.frame, tibble or data.table
#' @param treatment a string with treatment status column
#' @return A list: regression_tables = regression output of treatment against all covariates, 
#' F_test = table with the F tests of each regression
#' @examples 
#' data <-data.frame(x1 = rnorm(n = 100, mean = 100, sd = 15), x2= rnorm(n = 100, mean = 65), 
#' treat = rep(c(0,1,2,3,4), each = 20))
#' balance_regression(data = data, treatment = "treat")
#' @details This functions runs a Linear Probability model of each treatment group & control on all the 
#' columns in data. For instance, if treatment column has values of (0,1,2), balance_regression will run two 
#' models: 1) LPM(treatment(0,1)~X'b) and 2) LPM(treatment(0,2)~X'b). The value are the regression tables and 
#' details of the F_test of these models.
#' @export
#' @importFrom magrittr %>%
balance_regression <- function(data, treatment) { 
    
    estadistico<-NULL
    
    data <- data %>% dplyr::arrange(!!rlang::sym(treatment))
    
    valores_trat <- base::unique(dplyr::pull(data, !!rlang::sym(treatment)))
    
    trats<-valores_trat[2:base::length(valores_trat)]
    
    bal_tables<-purrr::map(trats, function (x)
        data %>%
            dplyr::filter(!!rlang::sym(treatment) == valores_trat[1] | !!rlang::sym(treatment) ==  !!x))
    
    formula<-glue::glue("{treatment}~ .")
    regressions <- purrr::map(bal_tables, ~stats::lm(as.formula(formula), data = .))
    
    nombres<-stringr::str_c("Msj", trats)
    
    names(regressions)<-nombres
    
    regression_tables<-purrr::map_dfc(regressions, function(x) broom::tidy(x))
    
    if (utils::packageVersion("dplyr") > "0.8.99.9002") {
        
    f<-purrr::map_dfc(regressions, function(x) base::summary(x)$fstatistic)
    p_values<-purrr::map_dbl(f, function(x) stats::pf(x[1], x[2], x[3], lower.tail = F))
    f_star <- purrr::map_dbl(f, function(x) stats::qf(p = 0.05, df1 = x[2], df2 = x[3]))
    
    f <- dplyr::bind_rows(f, f_star, p_values)
    
    r_cuadrada <- purrr::map_dbl(regressions, function(x) base::summary(x)$r.squared)
    
    f <-dplyr::bind_rows(f, r_cuadrada)
    
    f <-
        f %>%
        dplyr::mutate(estadistico = c("F-statistic", "k", "n-k-1", "F_critical", "p_value", "R cuadrada")) %>%
        dplyr::select(estadistico, dplyr::everything())
    
    
    objetos <- base::list("regression_tables" = regression_tables, "F_test" = f)
    
    return(objetos)
    
    } else {
    
        f<-purrr::map_dfr(regressions, function(x) base::summary(x)$fstatistic)
        p_values<-purrr::map_dbl(f, function(x) stats::pf(x[1], x[2], x[3], lower.tail = F))
        f_star <- purrr::map_dbl(f, function(x) stats::qf(p = 0.05, df1 = x[2], df2 = x[3]))
        
        f <- dplyr::bind_rows(f, f_star, p_values)
        
        r_cuadrada <- purrr::map_dbl(regressions, function(x) base::summary(x)$r.squared)
        
        f <-dplyr::bind_rows(f, r_cuadrada)
        
        f <-
            f %>%
            dplyr::mutate(estadistico = c("F-statistic", "k", "n-k-1", "F_critical", "p_value", "R cuadrada")) %>%
            dplyr::select(estadistico, dplyr::everything())
        
        
        objetos <- base::list("regression_tables" = regression_tables, "F_test" = f)
        
        return(objetos)
        
        
    }
    
    }







R/summary_statistics.R
#' summary_statistics() Creates summary statistics table of all numeric variables in data 
#' @param data A data.frame, tibble or data.table
#' @param probs The quantiles to compute. Default is c(0, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 1)
#' @param na.rm whether to exclude NA's from calculations
#' @return A tibble with the Mean, N (not NA) and probs selects for each numeric column
#' @examples 
#' data <-data.frame(x = c(1:5), y = c(100, 200, 300, 410, 540), z = rep("c", 5))
#' summary_statistics(data)
#' @details This function computes the selected quantiles, mean and N values of all the numeric columns of data. 

#' @export
#' @importFrom magrittr %>%
summary_statistics<- function(data, probs = c(0, 0.05, 0.1, 0.25, 0.5,
                                                   0.75, 0.9, 0.95, 1), na.rm = T) {

    base <- data %>% dplyr::ungroup()

    variables_numericas <- base %>% dplyr::select_if(is.numeric)
    
    statistic<-NULL

    estadisticas_descriptivas <- purrr::map_dfc(.x = variables_numericas,
                                                .f = function(x) stats::quantile(x, probs = probs, na.rm = na.rm))
    
    estadisticas_descriptivas$statistic<-base::as.character(probs) 
    
    estadisticas_descriptivas <-
        dplyr::select(estadisticas_descriptivas, statistic, dplyr::everything())

    medias <- purrr::map_dfc(.x = variables_numericas,
                             .f = ~mean(., na.rm = na.rm) )
    
    medias$statistic<-"mean"

    medias<-
        medias %>%
        dplyr::select(statistic, dplyr::everything())


    n_s <- purrr::map_dfc(.x = variables_numericas,
                             .f = ~sum(!is.na(.)) )
    
    n_s$statistic<-"n"
    
    n_s<-
        n_s %>%
        dplyr::select(statistic, dplyr::everything())
    
    estadisticas_descriptivas<-dplyr::bind_rows(medias, n_s ,estadisticas_descriptivas)
    
    estadisticas_descriptivas<-
        estadisticas_descriptivas %>% 
        tidyr::pivot_longer(-statistic, names_to = "variable") %>%
        tidyr::pivot_wider(names_from = "statistic")
    
    

    return(estadisticas_descriptivas)



}


